README
Site-Configuration

This is an experimental package to help Grid sites configure their systems more
easily. It currently reads configuration files in INI format and makes the data
therein available through tied hashes. The only extra functionality right now
is easy access to VO specific parameters.

This is still in the experimental stages and interfaces may change often and
without notification and/or backward compatibility. Be warned!

INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc Site::Configuration


LICENSE AND COPYRIGHT

Copyright (C) 2012 Stichting FOM

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    L<http://www.apache.org/licenses/LICENSE-2.0>

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.



Changes
Revision history for Site-Configuration

0.05    2013-04-25 09:06

	Restructuring of the VO package. Add support for
	writing configuration changes.

0.04    2013-04-23 23:22

	The default VO configuration directory is changed from
	/etc/vo-support to /etc/vo-support/vos, to match the
	structure of the vo-support package.

0.03    2012-06-26 14:46

	The modules have been extensively refactored to be properly
	Object Oriented. Documentation is updated accordingly.

	VO.pm is added to access VO configuration and FQANs.

0.02    2012-06-07 15:55

	readconfig() returns a reference to the bound hash instead of
	a copy, so changes may be written back.

0.01    2012-06-01 23:41

        First version, unreleased



lib/Site/Configuration.pm
# Experimental perl module
#  Copyright 2012  Stichting FOM
#
#   Licensed under the Apache License, Version 2.0 (the "License");
#   you may not use this file except in compliance with the License.
#   You may obtain a copy of the License at
#
#       http://www.apache.org/licenses/LICENSE-2.0
#
#   Unless required by applicable law or agreed to in writing, software
#   distributed under the License is distributed on an "AS IS" BASIS,
#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#   See the License for the specific language governing permissions and
#   limitations under the License.

package Site::Configuration;

use 5.006;
use strict;
use warnings;
use Config::IniFiles;
use Carp;

=head1 NAME

Site::Configuration - Access site-local configuration data

=head1 VERSION

Version 0.05

=cut

our $VERSION = "0.05";


sub new {
  my $class = shift;
  my $self = { CONFDIR => "/etc/siteinfo",
	       CONFSET => {},
	       ERRMSG => []
	     };

  if (@_) { $self->{CONFDIR} = shift }
  bless ($self, $class);
  return $self;
}


=head1 SYNOPSIS

    use Site::Configuration;
    my $siteconf = Site::Configuration->new();
    my $ce = $siteconf->readconfig("ce.conf");
    my $node = $ce->{CE};
    my $queues = $ce->{$node}{queues};
    $ce->{$node}{SMPsize} = 8;
    $siteconfig->writeconfig($ce);

=head1 DESCRIPTION

The Site::Configuration module offers an interface into the local
configuration data stored in Ini files. There is currently no
enforcement of the structure of any of the configuration files other
than a few basic formatting rules, so essentially it is now little
more than a thin shell around L<Config::IniFiles>.

The constructor accepts one optional argument, the name of a
configuration directory to use. The default value is
C</etc/site-info/>.

    my $siteconfig = Site::Configuration->new("/etc/alt-site");

The configuration directory may be read and set using the confdir() method:

    my $oldconfdir = $siteconfig->confdir();
    $siteconfig->confdir("/etc/another-site");

Reading a configuration file is triggered by the readconfig() method.

    my $ceconf = $siteconfig->readconfig("ce.conf");

The name that is passed should be the file name relative to the configuration
directory. The returned value is a reference to a tied hash.

Changing the configuration directory only affects reading new
configuration files; it does not cause existing files to be closed or
re-read. Site::Configuration keeps internal links to all the opened files
by their relative name.

=head2 Error handling

If an error occurs, the method readconfig() returns undef and the error
messages are stored in an internal array, which can be retrieved with $obj->errmsg().
Calling this method will clear the error, so be sure to store the values. Calling
readconfig() multiple times without clearing the error messages will cause them
to accumulate.

    if (!defined $ceconf) {
        print STDERR "$_\n" foreach $siteconfig->errmsg();
    }

=cut

# getter/setter
sub confdir {
  my $self = shift;
  if (@_) { $self->{CONFDIR} = shift }
  return $self->{CONFDIR};
}

# return and clear
sub errmsg {
  my $self = shift;
  my @ret = @{$self->{ERRMSG}};
  $self->{ERRMSG} = [];
  return @ret;
}

# Readconfig returns a reference to a hash tied to a Config::IniFiles
# object.
sub readconfig {
  my $self = shift;
  my $conf = shift;

  return $self->{CONFSET}->{$conf} if defined $self->{CONFSET}->{$conf};

  my $confdir = $self->{CONFDIR};
  # Sanity: does the configuration file exist?
  -f "$confdir/$conf" or do {
    push @{$self->{ERRMSG}}, sprintf "Missing configuration file %s.", "$confdir/$conf";
    return undef };

  tie my %ini, 'Config::IniFiles',
    ( -file => "$confdir/$conf",
      -fallback => "DEFAULT",
      -handle_trailing_comment => 1,
      -allowcontinue => 1,
      -nocase => 0) or
	do {
	  push @{$self->{ERRMSG}}, "There was an error reading configuration file $confdir/$conf:";
	  push @{$self->{ERRMSG}}, $_ foreach @Config::IniFiles::errors;
	  return undef;
	};
  $self->{CONFSET}->{$conf} = \%ini;
  return \%ini;
};

# Writing back the configuration. Argument is the tie reference
# as returned from readconfig

sub writeconfig {
  my $self = shift;
  my $iniref = shift;

  my $ini = tied ( %{$iniref} ) or
    croak "iniref is not a tied hash";
  if (!defined $ini->GetFileName()) {
    push @{$self->{ERRMSG}}, "Writeconfig: Inifile has no associated Filename.";
    return undef;
  }
  return $ini->RewriteConfig();
}

=head1 SEE ALSO

L<Config::IniFiles>

=head1 AUTHOR

Dennis van Dok <dennisvd at nikhef.nl>

Please report any bugs or feature requests to C<grid-mw-security-support at nikhef.nl>.

=head1 COPYRIGHT AND LICENSE

Copyright 2012 Stichting FOM

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    L<http://www.apache.org/licenses/LICENSE-2.0>

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.


=cut

1; # End of Site::Configuration


lib/Site/Configuration/VO.pm
# Experimental perl module
#  Copyright 2012  Stichting FOM
#
#   Licensed under the Apache License, Version 2.0 (the "License");
#   you may not use this file except in compliance with the License.
#   You may obtain a copy of the License at
#
#       http://www.apache.org/licenses/LICENSE-2.0
#
#   Unless required by applicable law or agreed to in writing, software
#   distributed under the License is distributed on an "AS IS" BASIS,
#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#   See the License for the specific language governing permissions and
#   limitations under the License.

package Site::Configuration::VO;

use 5.006;
use strict;
use warnings;
use Site::Configuration;
use Carp;

=head1 NAME

Site::Configuration::VO - Access Virtual Organisation configuration

=head1 VERSION

Version 0.05

=cut

our $VERSION = '0.05';

=head1 SYNOPSIS

    use Site::Configuration::VO;

    my $voconf = Site::Configuration::VO->new();
    my @fqans = $voconf->get_fqans(vo => "atlas");
    my $param = $voconf->get_vo_param( vo    => "atlas",
                                       param => "poolprefix",
                                       fqan  => "/atlas/Role=pilot" );
    $voconf->vo("atlas");
    $voconf->set_vo_param( param => "poolaccounts",
                           fqan  => "/atlas/Role=pilot",
                           value => "100" )

=cut


=head1 DESCRIPTION

The Site::Configuration::VO module provides read access to the
configuration of Virtual Organisations on the local system. Like
Site::Configuration, the configuration files are in Ini file format,
one file per VO. The default directory to look for configuration files
is /etc/vo-support/vos, but it can be overridden by using the confdir
object method.

VO configuration is organised in a single file named I<vo>.conf, with
sections for each FQAN. Settings that are global to the VO go in the
top section called [DEFAULT], but this section header may be
omitted. Any settings preceding the first FQAN is considered to be in
the [DEFAULT] section.

The constructor accepts a single optional argument, which is the directory
to read the VO configuration files from.

    my $voconf = Site::Configuration::VO->new();
    my $voconf = Site::Configuration::VO->new("/etc/othervos");

The configuration directory may be changed by using the confdir() method.

    my $oldconfig = $voconfig->confdir();
    $voconf->confdir("/etc/othervo");

Be aware that changing the directory after accessing a VO's parameters won't work
as Site::Configuration::VO keeps the handle to the original file.

The FQANs for a VO are enumerated with the get_fqans() method.

    my @fqans = $voconf->get_fqans(I<vo>);

Specific VO parameters are retrieved with the get_vo_param() method.

    my $defaultse = $voconf->get_vo_param(pvier, "defaultse")
    my $poolprefix = $voconf->get_vo_param(pvier, "poolprefix", "/pvier" )

The first form retrieves a VO global parameter. The second form looks up a
parameter in a specific FQAN section.

=head2 File Format

The VO configuration files are found in /etc/vo-support/I<vo>.conf, unless the
directory was changed as described above. The files are in Ini file format,
see L<Config::IniFiles>.

    # example configuration file for pvier
    SoftwareDir = /data/esia/pvier
    DefaultSE = tbn18.nikhef.nl

    [/pvier]
    poolaccounts = 30
    poolprefix = pvier
    groupmapping = pvier

    [/pvier/Role=lcgadmin]
    poolprefix = pvsgm
    poolaccounts = 10


=head2 Error Handling

In case of errors, get_fqans() returns an empty list and
get_vo_param() returns undef.  The array of error messages may be
retrieved using the errmsg() method.

      print STDERR "$_\n" foreach $voconf->errmsg();

Any errors that occur on consecutive calls will accumulate in this array;
the errmsg() method clears the array again.

=head1 METHODS

Here follows the detailed documentation of the public methods.

=head2 new (constructor)

The constructor can be called with two named parameters: B<confdir> for
the configuration directory to use (defaults to F</etc/vo-support/vos>)
and B<vo> to indicate which VO to use in subsequent calls to other methods.
Both can be set and reset at any time.
Example:
    my $voconfig = Site::Configuration::VO->new(vo => "atlas")

=cut

sub new {
  my $class = shift;
  my %params = @_;
  my $confdir = $params{confdir} // "/etc/vo-support/vos";
  my $self = { SITECONF => Site::Configuration->new($confdir),
	       VO => $params{vo},
	       VOCONF => {},
	       ERRMSG => [],  # array of accumulated error messages
	     };
  bless ($self, $class);
  return $self;
}


# return and clear
sub errmsg {
  my $self = shift;
  my @ret = @{$self->{ERRMSG}};
  $self->{ERRMSG} = [];
  return @ret;
}

=head2 accessors: vo and confdir

Two accessor methods allow getting and setting of the VO
and configuration directory.
    $voconfig->vo();        # returns current VO
    $voconfig->vo("atlas"); # sets current VO

    $voconfig->confdir();       # returns current configuration directory
    $voconfig->confdir("/tmp"); # sets current configuration directory

=cut

# getter/setter
sub vo {
  my $self = shift;
  if (@_) { $self->{VO} = shift }
  return $self->{VO};
}

sub confdir {
  my $self = shift;
  return $self->{SITECONF}->confdir(@_); # transparent?
}

=head2 get_vo_param

This method retrieves parameters from the VO configuration file.
The parameters that may be passed are:

=over

=item B<vo>

The VO of which to retrieve the configuration.

=item B<fqan>

The FQAN section in the config file to retrieve
the parameter from. If this is not specified, the [DEFAULT]
section is used (if no [DEFAULT] section is present, this
refers to all parameters in the file before the first FQAN
section).

=item B<param> (mandatory)

The name of the parameter to retrieve.

=back

=head3 Return Value

=over

=item The requested parameter

In case the parameter was found in the file

=item undef

This can happen in three situations:

=over

=item The VO wasn't specified, and no VO was ever set in the object.

=item The configuration file for the VO could not be read.

=item The requested parameter was missing in the configuration file.

=back

The only way to distinguish between these cases is by calling
the errmsg() method.

=back

=cut

sub get_vo_param {
  my $self = shift;
  my %params = @_;
  my $vo = $params{'vo'} // $self->{VO} or
    croak "get_vo_param: VO not set.";

  my $fqan = $params{fqan} // "DEFAULT";

  my $param = $params{param};
  if (!defined $param) {
      croak "get_vo_param: missing 'param' argument";
  }
  # check if we already have the VO
  $self->_readvo($vo) or return undef;
  # look up the parameter
  my $val = $self->{VOCONF}->{$vo}->{$fqan}{$param};
  return $val;
}

=head2 set_vo_param

The counterpart of retrieving VO parameters is this method which sets
them. This immediately results in the rewriting of the configuration
file. The following parameters may be passed:

=over

=item vo

=item fqan

=item param

=item value

=back

=head3 Return Value

Returns undef in case of failure, 1 otherwise.

=cut

# This may be dangerous; going to try and rewrite the original config file!
# TODO: this code is so much like get_vo_param it should be refactored
# into a single module.

sub set_vo_param {
  my $self = shift;
  my %params = @_;
  my $vo = $params{vo} // $self->{VO};
  my $fqan = $params{fqan} // "DEFAULT";
  foreach $_ (qw(param value)) {
    if ( ! defined $params{$_} ) {
      croak "Missing parameter '$_' to function set_vo_param";
    }
  }
  my ($param, $value) = ($params{param}, $params{value});

  $self->_readvo($vo) or return undef;
  if ( ! defined $self->{VOCONF}->{$vo}) {
    push @{$self->{ERRMSG}}, "Cannot set VO parameter.";
    return undef;
  }
  $self->{VOCONF}->{$vo}->{$fqan}{$param} = $value;
  if (! defined $self->{SITECONF}->writeconfig($self->{VOCONF}->{$vo}) ) {
    push @{$self->{ERRMSG}}, $_ foreach $self->{SITECONF}->errmsg();
    push @{$self->{ERRMSG}}, "Cannot write configuration for VO '$vo'.";
    return undef;
  }
  return 1;
}

=head2 get_fqans

Returns a list of FQANs configured for the VO; this is the list of
sections in the configuration file that start with a '/' and a
letter. Other sections are ignored.

This method takes a single optional parameter to specify the VO.
    my @fqans = $voconfig->get_fqans(vo => "atlas");

=cut

sub get_fqans {
  my $self = shift;
  my %params = @_;
  my $vo = $params{vo} // $self->{VO};
  if ( ! defined $vo ) {
    push @{$self->{ERRMSG}}, "get_fqans: VO not specified";
    return undef;
  }

  my @fqans = (); # return value.
  $self->_readvo($vo) or return undef;
  for my $section (keys %{$self->{VOCONF}->{$vo}}) {
    if ($section =~ m{^/[[:alpha:]]}) {
      push @fqans, $section
    }
  }
  return @fqans;
}

=head2 isenabled

Arguments: I<vo>.

Returns 1 if the VO is enabled, 0 otherwise. If the I<enabled> parameter
is not found in the configuration file, the default value is assumed to be
'true'.

Returns undef if the VO is not configured.

=cut

sub isenabled {
  my $self = shift;
  my %args = @_;
  my $vo = $args{vo} // $self->{VO};
  if (! defined $vo) { croak "isenabled: no VO given" };
  $self->_readvo($vo) or return undef;
  my $enabled = $self->get_vo_param(vo => $vo, param => "enabled");
  return (! defined $enabled) || $enabled !~ m/(false|no|disabled|0)/i;
}

=head2 enable, disable

These methods set the I<enabled> parameter in the [DEFAULT] section of the
configuration to 'true' and 'false', respectively.

    $voconfig->enable( vo => "atlas" )

Some care is taken to prevent unnecessary rewrites of the configuration; if the VO is already in
the requested state nothing is done.

The return value is 1 on success, 0 otherwise.

If there is no configuration file, no action is taken and undef is returned.

=cut

sub enable {
  my $self = shift;
  my %args = @_;
  my $vo = $args{vo} // $self->{VO} or croak "enable: no VO given";
  return 1 if $self->isenabled(vo => $vo);
  return $self->set_vo_param(vo => $vo, param => "enabled", value => "true");
}

sub disable {
  my $self = shift;
  my %args = @_;
  my $vo = $args{vo} // $self->{VO} or croak "disable: no VO given";
  return 1 if ! $self->isenabled(vo => $vo);
  return $self->set_vo_param(vo => $vo, param => "enabled", value => "false");
}

# This is a 'private' method

sub _readvo {
  my $self = shift;
  my $vo = shift;
  # read vo configuration only once
  if ( ! defined $self->{VOCONF}->{$vo}) {
    $self->{VOCONF}->{$vo} = $self->{SITECONF}->readconfig("vos/" . $vo . ".conf");
  }
  if ( ! defined $self->{VOCONF}->{$vo}) {
    push @{$self->{ERRMSG}}, $_ foreach $self->{SITECONF}->errmsg();
    push @{$self->{ERRMSG}}, "Cannot read configuration for VO '$vo'.";
    return undef;
  }
  return $self->{VOCONF}->{$vo};
}

=head1 FILES

/etc/vo-support/vos/*.conf

=head1 SEE ALSO

L<Site::Configuration>, L<Config::IniFiles>

=head1 AUTHOR

Dennis van Dok <dennisvd@nikhef.nl>

Please report any bugs or feature requests to <grid-mw-security-support@nikhef.nl>.

=head1 COPYRIGHT AND LICENSE

Copyright 2012 Stichting FOM

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

L<http://www.apache.org/licenses/LICENSE-2.0>

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.


=cut


1; # End of Site::Configuration::VO
